idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/editor/folding/FoldingSyntax.kt

```diff
@@ -1,48 +1,34 @@
 package arrow.meta.ide.dsl.editor.folding
 
 import arrow.meta.ide.IdeMetaPlugin
+import arrow.meta.ide.phases.editor.extension.ExtensionProvider
 import arrow.meta.phases.ExtensionPhase
 import com.intellij.lang.ASTNode
 import com.intellij.lang.Language
 import com.intellij.lang.folding.FoldingBuilder
-import com.intellij.lang.folding.FoldingBuilderEx
 import com.intellij.lang.folding.FoldingDescriptor
-import com.intellij.lang.folding.LanguageFolding
 import com.intellij.openapi.editor.Document
-import com.intellij.psi.PsiElement
-import com.intellij.util.KeyedLazyInstance
 import org.jetbrains.kotlin.idea.KotlinLanguage
 
 interface FoldingSyntax {
-  fun IdeMetaPlugin.addFoldingBuilder(
+
+  fun IdeMetaPlugin.registerFoldingBuilder(foldingBuilder: FoldingBuilder, lang: Language = KotlinLanguage.INSTANCE): ExtensionPhase =
+    ExtensionProvider.AddFoldingBuilder(lang, foldingBuilder)
+
+  fun FoldingSyntax.foldingBuilder(
     placeHolderText: (node: ASTNode) -> String?,
-    foldRegions: (root: PsiElement, document: Document, quick: Boolean) -> List<FoldingDescriptor>,
+    foldRegions: (node: ASTNode, document: Document) -> Array<FoldingDescriptor>,
```

Comment:
```suggestion
    foldRegions: (node: ASTNode, document: Document) -> List<FoldingDescriptor>,
```

Reply:
Hmmm the interface is defined like that:
```
  @NotNull
  FoldingDescriptor[] buildFoldRegions(@NotNull ASTNode node, @NotNull Document document);
```

Reply:
Although I can do this and reads better.. if its what u mean:
```
      override fun buildFoldRegions(node: ASTNode, document: Document): Array<FoldingDescriptor> =
        foldRegions(node.psi as KtElement, document).toTypedArray()
```

Reply:
yeah, looks good 👍